File: 1036-rotting-oranges/1036-rotting-oranges.java
class Solution {
    public int orangesRotting(int[][] grid) {
        //define a queue which will contain an array of integer each storing grid position where oranges are rotten 
        Queue<int[]> queue = new LinkedList<>();
        //define a fresh pointer to track total number of fresh oranges
        int fresh = 0;
        // define a time pointer to identity total time taken
        int time = 0;
        //loop through the grid, if an orange is rotten push it onto the queue and if it is fresh increase counter for fresh
        int m = grid.length;
        int n = grid[0].length;
        for(int row=0; row<m;row++){
            for(int col=0;col<n;col++){
                int cell = grid[row][col];
                if(cell==1){
                    fresh++;
                }
                if(cell==2){
                    queue.offer(new int[]{row, col});
                }
            }
        }
        int[][] directions = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
        //until no fresh oranges are left or until queue is empty do bfs where oranges are rotten
        while(fresh>0 && !queue.isEmpty()){
            int length = queue.size();
            for(int i=0; i<length; i++){
                int[] cell = queue.poll();
                int row = cell[0];
                int col = cell[1]; 
                //loop through each direction to get adjacent cells if they are fresh make them rotten and add it to queue
                for(int[] direction: directions){
                    int r = row + direction[0];
                    int c = col + direction[1];
                    if(r>=0 && r<m && c>=0 && c<n && grid[r][c]==1){
                        grid[r][c] = 2;
                        fresh--;
                        queue.offer(new int[]{r,c});
                    }
                }
            }
            //increase time pointer after each cycle
            time++;
        }
        //if fresh pointer is not zero return -1 else return time value
        return fresh>0 ? -1 : time;
    }
}

File: 1036-rotting-oranges/1036-rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])
        q = collections.deque()
        fresh = 0
        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == 2:
                    q.append((r, c))
                if grid[r][c] == 1:
                    fresh += 1
        directions = [[0, 1], [1, 0], [0, -1], [-1, 0]]
        time = 0
        while fresh > 0 and q:
            for _ in range(len(q)):
                r, c = q.popleft()
                for dr, dc  in directions:
                    nr, nc = r + dr, c + dc
                    if nr in range(ROWS) and nc in range(COLS) and grid[nr][nc] == 1:
                        grid[nr][nc] = 2
                        fresh -= 1
                        q.append((nr, nc))
            time += 1
        return time if fresh == 0 else -1

